### DISCUSSION TITLE
Empthy CFD Volume Plot

### DISCUSSION THREAD

**Discussion by SantiagoPendon** [pyvista/pyvista#5208] [Q&A]
Hi to everyone. I'm pretty new in pyVista and their way to work. We are trying to reproduce the example posted in:
https://docs.pyvista.org/version/stable/examples/99-advanced/openfoam-tubes.html#sphx-glr-examples-99-advanced-openfoam-tubes-py
 To be clear, we are trying to repeat the last example, when it makes a volume representation.  The others have been done correctly.

The expected result is 
![Captura de pantalla 2023-11-21 163759](https://github.com/pyvista/pyvista/assets/151015554/3fafa2b4-cc9f-4ca7-b27a-ccbf37185278)

But we obtain: 

![Captura de pantalla 2023-11-21 163850](https://github.com/pyvista/pyvista/assets/151015554/62b84304-d71a-4331-be32-0dd8a30566e6)

With our own data the same result is obtained so I suspect is related to versions or incompatibilities.

My versions are:

-pyvista:  0.42.3
-vtk: 9.2.6

Thanks a lot.

**Comment by MatthewFlamm** [comment]
Can you share what comes out of `pv.Report()`?  In my experience, depending on hardware and OS, you might have to use different settings to get volume rendering working correctly.  Others may have experience with specific settings to try adjusting once we know your environment.

**Reply by SantiagoPendon** [reply]
Hi Matthew, I'm thankful for your help. If I snip the code that I use, you will see:
![Captura de pantalla 2023-11-22 083332](https://github.com/pyvista/pyvista/assets/151015554/7899fcbd-fcd6-45ab-be63-c338cc7e2306)

Then, the command pv.Report() produces: 
![Captura de pantalla 2023-11-22 083818](https://github.com/pyvista/pyvista/assets/151015554/7e4a439f-b8f7-46cd-b6f1-a4131ee5478b)

Also, Im running this in Spyder, with Python´s version 3.11.4

**Comment by jmb64** [comment]
Hi,
 
    I think I may have encountered the same problem as Santiago here with the pyvista example (https://docs.pyvista.org/examples/99-advanced/openfoam-tubes.html).  If my interpretation is correct the empty volume is what you get when reading model output at timestep 0.
The 'U' field is identically 0 in the file.

I changed the original example code loading block 0:

block = examples.download_openfoam_tubes()
air = block[0]

by the following:

filename = examples.download_openfoam_tubes(load=False)
reader = pv.POpenFOAMReader(filename)
print(f"Available Time Values: {reader.time_values}")
#gives ---> Available Time Values: [0.0, 1000.0]            (2 timesteps available)
reader.set_active_time_value(1000.0)
mesh = reader.read()
air = mesh[0]

Skipping timestep 0 and loading timestep 1000, I got the expected output:
![pasted_image](https://github.com/user-attachments/assets/348b3a16-0b89-4adc-a699-9ccdd1b0b6df)



For reference, here is the openfoam output file (Turbo-Incompressible_3-Run_1-SOLUTION_FIELDS.zip) after unzipping:


ls -lRt  openfoam_fvm/  
openfoam_fvm/:
total 37348
drwxr-xr-x 3 jmb001 eccc_rpnenv     4096 Sep 29 19:48 1000
drwxr-xr-x 2 jmb001 eccc_rpnenv     4096 Sep 29 19:48 0
drwxr-xr-x 2 jmb001 eccc_rpnenv     4096 Sep 28 20:19 system
drwxr-xr-x 3 jmb001 eccc_rpnenv     4096 Sep 28 20:19 constant
-rw-r--r-- 1 jmb001 eccc_rpnenv 38228176 Sep 28 20:19 Turbo-Incompressible_3-Run_1-SOLUTION_FIELDS.zip
-rw-r--r-- 1 jmb001 eccc_rpnenv        0 Apr  4  2023 case.foam

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/1000:
total 59665
drwxr-xr-x 3 jmb001 eccc_rpnenv     4096 Sep 28 20:19 uniform
-rw-r--r-- 1 jmb001 eccc_rpnenv 18669347 Apr  4  2023 U
-rw-r--r-- 1 jmb001 eccc_rpnenv  6317068 Apr  4  2023 k
-rw-r--r-- 1 jmb001 eccc_rpnenv  7623022 Apr  4  2023 nut
-rw-r--r-- 1 jmb001 eccc_rpnenv  5933432 Apr  4  2023 omega
-rw-r--r-- 1 jmb001 eccc_rpnenv  5586223 Apr  4  2023 p
-rw-r--r-- 1 jmb001 eccc_rpnenv 16912102 Apr  4  2023 phi

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/1000/uniform:
total 2
drwxr-xr-x 2 jmb001 eccc_rpnenv 4096 Sep 28 20:19 functionObjects
-rw-r--r-- 1 jmb001 eccc_rpnenv  367 Apr  4  2023 cumulativeContErr.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv 1014 Apr  4  2023 time

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/1000/uniform/functionObjects:
total 1
-rw-r--r-- 1 jmb001 eccc_rpnenv 330 Apr  4  2023 functionObjectProperties.gz

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/0:
total 3
-rw-r--r-- 1 jmb001 eccc_rpnenv 2076 Apr  4  2023 U
-rw-r--r-- 1 jmb001 eccc_rpnenv 2131 Apr  4  2023 k
-rw-r--r-- 1 jmb001 eccc_rpnenv 1950 Apr  4  2023 nut
-rw-r--r-- 1 jmb001 eccc_rpnenv 2121 Apr  4  2023 omega
-rw-r--r-- 1 jmb001 eccc_rpnenv 1610 Apr  4  2023 p

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/system:
total 5
-rw-r--r-- 1 jmb001 eccc_rpnenv  376 Apr  4  2023 changeDictionaryDict
-rw-r--r-- 1 jmb001 eccc_rpnenv 1161 Apr  4  2023 controlDict
-rw-r--r-- 1 jmb001 eccc_rpnenv  136 Apr  4  2023 decomposeParDict
-rw-r--r-- 1 jmb001 eccc_rpnenv 1060 Apr  4  2023 fieldOperationsDict.continuedrun.encode
-rw-r--r-- 1 jmb001 eccc_rpnenv 1320 Apr  4  2023 fieldOperationsDict.decode
-rw-r--r-- 1 jmb001 eccc_rpnenv 1047 Apr  4  2023 fieldOperationsDict.encode
-rw-r--r-- 1 jmb001 eccc_rpnenv   92 Apr  4  2023 fvOptions
-rw-r--r-- 1 jmb001 eccc_rpnenv  918 Apr  4  2023 fvSchemes
-rw-r--r-- 1 jmb001 eccc_rpnenv 1056 Apr  4  2023 fvSolution

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/constant:
total 3
drwxr-xr-x 2 jmb001 eccc_rpnenv 4096 Sep 28 20:19 polyMesh
-rw-r--r-- 1 jmb001 eccc_rpnenv  176 Apr  4  2023 transportProperties
-rw-r--r-- 1 jmb001 eccc_rpnenv  194 Apr  4  2023 turbulenceProperties
-rw-r--r-- 1 jmb001 eccc_rpnenv   96 Apr  4  2023 RASProperties

/home/jmb001/data/ppp5/PY/PYVISTA/openfoam_fvm/constant/polyMesh:
total 11410
-rw-r--r-- 1 jmb001 eccc_rpnenv 1711347 Apr  4  2023 neighbour.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv  773620 Apr  4  2023 owner.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv     328 Apr  4  2023 pointZones.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv 2260004 Apr  4  2023 points.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv    2297 Apr  4  2023 boundary
-rw-r--r-- 1 jmb001 eccc_rpnenv     328 Apr  4  2023 cellZones.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv     328 Apr  4  2023 faceZones.gz
-rw-r--r-- 1 jmb001 eccc_rpnenv 6900439 Apr  4  2023 faces.gz




pyvista.Report()
--------------------------------------------------------------------------------
  Date: Mon Sep 30 16:18:03 2024 UTC

                OS : Linux (Ubuntu 22.04)
            CPU(s) : 80
           Machine : x86_64
      Architecture : 64bit
               RAM : 992.2 GiB
       Environment : Python
       File system : xfs
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : NVIDIA A100 80GB PCIe/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 525.85.12
  MathText Support : True

  Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC
  12.3.0]

           pyvista : 0.44.0
               vtk : 9.2.6
             numpy : 1.26.4
        matplotlib : 3.8.4
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 10.3.0
           imageio : 2.34.2
         pyvistaqt : 0.11.1
             PyQt5 : 5.15.9
           IPython : 8.26.0
          colorcet : 3.1.0
           cmocean : 4.0.3
        ipywidgets : 8.1.3
             scipy : 1.14.0
              tqdm : 4.66.4
        jupyterlab : 4.2.3
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------

BTW, I no nothing about openfoam fle format ... 

Jean-Marc

**Reply by MatthewFlamm** [reply]
There were some regressions in this dataset, see https://github.com/pyvista/pyvista/pull/6380/, for pyvista 0.44.0.  I believe 0.44.1 has the fix, which is more or less exactly what you have in this comment.

I still suspect the original discussion issue is related to volume rendering rather than the dataset, but I'm not sure either way.

**Reply by jmb64** [reply]
Hi Matthew,

Thanks for letting me know. Not familiar enough to argue here :-)
Recently been experimenting with pyvista, I'm trying to figure out if I can do direct volume rendering 
of an UnstructuredGrid (essentially gridpoints from an ocean model with a curvilinear grid)    
without hitting the performance issue stated here (https://github.com/pyvista/pyvista/discussions/5035) 
that would make it unbearable ...

Best Regards,
 
Jean-Marc

**Reply by MatthewFlamm** [reply]
I find that rendering several ISO contour surfaces with transparency can give a decent effect of 3D space.

**Reply by jmb64** [reply]
Yes, good point.   I guess I'm trying to see where pyvista (available algorithms) stands w/r to state-of-the art in the field ?
What can be done and what can't or maybe get a feel for the problem size (grid size) that's bearable ...

(3D) Visualization is not my field of expertise and I'm trying to learn a thing or two  ... :-). It seems from browsing the litterature 
that progress is being made in this area (volume rendering unstructured grids). Whether this translates into actual implementation, 
I don't know. Maybe in Paraview, Vapor or other applications. I understand that pyvista  relies on    
"P. Bunyk, A. Kaufman, and C. T. Silva. Simple, Fast, and Robust Ray Casting of Irregular Grids." 
throug vtk's RayCastFunction: vtkUnstructuredGridBunykRayCastFunction for this case.  
Feels like venturing on slippery ground ... hihihi!  :-)

Cheers,
Jean-Marc

**Comment by SantiagoPendon** [comment]
Unsubscribe

**Reply by MatthewFlamm** [reply]
There is a button on this discussion that you can use to unsubscribe.  Mine looks like this 
![image](https://github.com/user-attachments/assets/376e9254-2f82-48aa-90fd-53a48c715b68)
